[patch] Probes Section cleanup (#154)
chipsalliance/firrtl-spec: spec.md

## Types

There are two probe types, `Probe<T>`{.firrtl} is a read-only variant and `RWProbe<T>`{.firrtl} is a read-write variant.
(See [@sec:probe-types]).

Examples:

``` firrtl
Probe<UInt<8>>
RWProbe<UInt<8>>
```

A `RWProbe<T>`{.firrtl} will be implicitly cast to a `Probe<T>`{.firrtl} whenever an expression of type `RWProbe<T>`{.firrtl} is assigned to a port with type `Probe<T>`{.firrtl}.

While `RWProbe<T>`{.firrtl} is strictly more powerful, `Probe<T>`{.firrtl} allows for more aggressive optimization.
You should consider using `Probe`{.firrtl} whenever you know you don't need to ability to force (see [@sec:forcing]).

## The `probe`{.firrtl} and `rwprobe`{.firrtl} Expressions

Given a reference to a circuit component, you can capture it in a probe with `probe`{.firrtl} for a read-only probe or with `rwprobe`{.firrtl} for a read-writeable probe.

## The `define`{.firrtl} Statement

Since neither `Probe<T>`{.firrtl} nor `RWProbe<T>`{.firrtl} is a connectable type ([@sec:connectable-types]), circuit components with these types cannot be connected to.
Instead, we use the `define`{.firrtl} statement to route probes through a FIRRTL circuit.

> The `define`{.firrtl} statements works similarly to `connect`{.firrtl}, but it has special interactions with conditional statements ([@sec:conditionals]).

Reviewer: ```suggestion
The `define`{.firrtl} statement works similarly to `connect`{.firrtl}, but it has special interactions with conditional statements ([@sec:conditionals]).
```

> `define`{.firrtl} statements may appear within conditional blocks, and when they do, they may connect circuit components defined in a local scope.

Reviewer: ```suggestion
`define`{.firrtl} statements may appear within conditional blocks, and when they do, they may refer to circuit components defined in a local scope.
```
(mostly, probably should avoid saying `define` "connects" anything? And I think idea here is answering the question "why would you use define from within a conditional?" and the answer is "so the things you're referring to are in-scope" (right?))

> However, `define`{.firrtl} are not subject to last connect semantics, and the `define`{.firrtl} statement is always and unconditionally in effect.

Reviewer: ```suggestion
However, `define`{.firrtl} statements are not subject to last connect semantics, and the `define`{.firrtl} statement is always and unconditionally in effect.
```
(feels awkward to use "define" as plural?)

  define z[1] = probe(w)
```

## The `read`{.firrtl} Expressions

Both `Probe`{.firrtl} and `RWProbe`{.firrtl} may be read from using the `read`{.firrtl} expression.

> Keep in mind that probes are intended for verification and isn't intended to be synthesizable.

Reviewer 2: ```suggestion
Keep in mind that probes are intended for verification and aren't intended to be synthesizable.
```

> A `read`{.firrtl} expression can reach arbitrarily far into the module hierarchy to read from a distant circuit component, which is arguably unphysical at worst and potentially expensive at best.

Reviewer 2: all of this feels very... commentary-like. I'm not sure what it really has to do with the firrtl spec , especially without considering the one ABI.

Author: Agreed. However, I think it's good to have this as a placeholder until we can formalize the distinction between "synthesizable"/"hardware" and "verification-only"/"simulation-only".

## Forcing

A `RWProbe`{.firrtl} may be forced using the `force`{.firrtl} and `force_initial`{.firrtl} commands.
When you force a component, it will ignore all other drivers and take on the value supplied by the force statement instead.
The `release`{.firrtl} and `release_initial`{.firrtl} statements end the forcing, reconnecting the circuit component to its usual driver.

> Like `read`{.firrtl}, the force statements are verification constructs.

Reviewer 2: what is a "verification construct". Is this a term we've defined? It's rather confusing given we have a bunch of other verification like constructs such as "printf" and "assert" -- how are these things like eachother?

Author: I am moving towards a formal definition of these things.

We'll get there 👍

  define z[1] = probe(w)
```

## The `read`{.firrtl} Expressions

Both `Probe`{.firrtl} and `RWProbe`{.firrtl} may be read from using the `read`{.firrtl} expression.

Keep in mind that probes are intended for verification and aren't intended to be synthesizable.
A `read`{.firrtl} expression can reach arbitrarily far into the module hierarchy to read from a distant circuit component, which is arguably unphysical at worst and potentially expensive at best.

## Forcing

A `RWProbe`{.firrtl} may be forced using the `force`{.firrtl} and `force_initial`{.firrtl} commands.
When you force a component, it will ignore all other drivers and take on the value supplied by the force statement instead.
The `release`{.firrtl} and `release_initial`{.firrtl} statements end the forcing, reconnecting the circuit component to its usual driver.

> Like `read`{.firrtl}, the force statements are verification constructs.

Reviewer: Just checking, "force" is not code-formatted because it's referring to the force "family" of statements?

Author: Yeah.

  connect y, p
```

## Probes and Layers

`Probe`{.firrtl} and `RWProbe`{.firrtl} types may be associated with a layer (see [@sec:layers]).

> When associated with a layer, the reference type may only be driven from tha layer.

Reviewer 2: ```suggestion
When associated with a layer, the reference type may only be driven from that layer.
```

For example:

``` firrtl
Probe<UInt<8>, A.B>     ; A.B is a layer
RWProbe<UInt<8>, A.B>
```

Probes are generally lowered to hierarchical names in Verilog.

> For details, see the FIRRTL ABI Specification.

Reviewer 2: (i see the same sentence repeated just below)

## Limitations

### `RWProbe`{.firrtl} with Aggregate Types

TODO.

For more information on the how probes are lowered, see the FIRRTL ABI Specification.

### No `RWProbe`{.firrtl}s of Ports on Public Modules

Conceptually, in order to force a value onto a circuit component, you need to be able to disconnect it from its former driver.
However, the driver of a port is another port in some other module.
And so, to force a port, you must have control over both "ends" of a port.

Trouble arises with ports on a public module (see TODO).
Since we don't control the remote side of ports (since they might live in another FIRRTL file), we cannot force these ports either.

For this reason, it is an error to use `rwprobe`{.firrtl} on any port on a public module.

### Input Ports

FIRRTL allows probes to be passed through ports.
However, due to the limitations of Verilog, there is an important restriction on where they may be used.

A module which only sends probes up towards its parent via output ports have no restriction.

> However, when a module receives a probe from its parent through an input port, you may not use it in a `read`{.firrtl} expression nor may you force it if the circuit component it refences does not live in a module that is a descendent of the current module.

Reviewer 2: is this saying that it's ok for a module `M` to have an input probe port, as long as outside of `M` that input probe is actually driven by an output probe port of `M` itself (either directly or from some child `C` of `M` where C has the output probe port?)

> However, when a module receives a probe from its parent through an input port, you may not use it in a `read`{.firrtl} expression nor may you force it if the circuit component it refences does not live in a module that is a descendent of the current module.

Reviewer: Yes that's the current rule (and also that what the probe at-point-of-use-in-a-read-or-force resolves to must be same in all instantiation contexts...).  The n-turn part of the de-squiggling was implemented but not landed yet in favor of focusing on what we practically would use (u-turn de-squiggling), at least in theory.  This kept things simpler, at least until alternate choice re-introduces the need for proving they're "equivalent" in all possible bottom-up cases.

This sort of thing: https://github.com/dtzSiFive/reftype-inputs/blob/main/ref_in_eq.fir (or more complicated: https://github.com/dtzSiFive/reftype-inputs/blob/main/ref_in_eq_subs_parent.fir )

If this PR is meant to keep things the same but clean them up, that's the current rule yes.  We may want to change it however.

For more information on the how probes are lowered, see the FIRRTL ABI Specification.

### External Modules

TODO

> Probe types may be specified as part of an external module (see [@sec:externally-defined-modules]), with the resolved referent for each specified using `ref`{.firrtl} statements.

Reviewer 2: ```suggestion
Probe type ports may be specified on the interface of an external module (see [@sec:externally-defined-modules]), with the resolved referent for each specified using `ref`{.firrtl} statements.
```

  connect y, p
```

## Probes and Layers

`Probe`{.firrtl} and `RWProbe`{.firrtl} types may be associated with a layer (see [@sec:layers]).

> When associated with a layer, the reference type may only be driven from that layer.

Reviewer: ```suggestion
When associated with a layer, the probe type may only be driven from that layer.
```
?

For example:

``` firrtl
Probe<UInt<8>, A.B>     ; A.B is a layer
RWProbe<UInt<8>, A.B>
```

For details on how probes are lowered, see the FIRRTL ABI Specification.

## External Modules

Probe types may be specified as part of an external module (see [@sec:externally-defined-modules]), with the resolved referent for each specified using `ref`{.firrtl} statements.

## Limitations

### `RWProbe`{.firrtl} with Aggregate Types

> Probes of aggregates do not working under all ABI versions.

Reviewer: ```suggestion
Probes of aggregates do not work under all ABI versions.
```

In these cases, it is recommended that if you have a need to probe a bundle, you instead use a bundle of probes, each targeting an individual subcomponent.

For more information on the how probes are lowered, see the FIRRTL ABI Specification.

### No `RWProbe`{.firrtl}s of Ports on Public Modules

Conceptually, in order to force a value onto a circuit component, you need to be able to disconnect it from its former driver.

> However, the driver of a port is another port in some other module.

Reviewer: This language is a bit off, LMK if not clear why or how to capture what you mean :+1: .

And so, to force a port, you must have control over both "ends" of a port.

Trouble arises with ports on a public module ([@sec:public-modules]).
Since we don't control the remote side of ports (since they might live in another FIRRTL file), we cannot force these ports either.

> For this reason, it is an error to use `rwprobe`{.firrtl} on any port on a public module.

Reviewer: Up to you, but at least to me this is same issue for public and extmodule's (and what used to be the main module/top)-- (the module definition, if FIRRTL-provided, must be public for it to be used as extmodule elsewhere).  Anyway I think this section could be tidied a little, I'll revisit with more concrete suggestions or we can work on it together if you'd like.

Reviewer: FWIW neither the other side of a public of extmodule necessarily is defined in FIRRTL.
